instance/models/instance.py

```diff
@@ -0,0 +1,436 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance app models - Instance
+"""
+
+# Imports #####################################################################
+
+import os
+
+from django.conf import settings
+from django.core.validators import RegexValidator
+from django.db import models
+from django.template import loader
+from django.utils import timezone
+from django_extensions.db.models import TimeStampedModel
+
+from instance import ansible, github
+from instance.gandi import GandiAPI
+from instance.github import fork_name2tuple, get_username_list_from_team
+from instance.log_exception import log_exception
+from instance.repo import clone_configuration_repo
+from instance.models.logging_mixin import LoggerInstanceMixin
+from instance.models.utils import ValidateModelMixin
+
+
+# Constants ###################################################################
+
+PROTOCOL_CHOICES = (
+    ('http', 'HTTP - Unencrypted clear text'),
+    ('https', 'HTTPS - Encrypted'),
+)
+
+gandi = GandiAPI()
+
+
+# Exceptions ##################################################################
+
+class InconsistentInstanceState(Exception):
+    """
+    Indicates that the status of an instance can't be determined
+    """
+    pass
+
+
+# Validators ##################################################################
+
+sha1_validator = RegexValidator(regex='^[0-9a-f]{40}$', message='Full SHA1 hash required')
+
+
+# Models ######################################################################
+
+class Instance(ValidateModelMixin, TimeStampedModel):
+    """
+    Instance - Group of servers running an application made of multiple services
+    """
+    sub_domain = models.CharField(max_length=50)
+    email = models.EmailField(default='contact@example.com')
+    name = models.CharField(max_length=250)
+
+    base_domain = models.CharField(max_length=50, default=settings.INSTANCES_BASE_DOMAIN)
+    protocol = models.CharField(max_length=5, default='http', choices=PROTOCOL_CHOICES)
+
+    last_provisioning_started = models.DateTimeField(blank=True, null=True)
+
+    class Meta:
+        abstract = True
+        unique_together = ('base_domain', 'sub_domain')
+
+    def __str__(self):
+        return '{0.name} ({0.url})'.format(self)
+
+    @property
+    def domain(self):
+        """
+        Instance domain name
+        """
+        return '{0.sub_domain}.{0.base_domain}'.format(self)
+
+    @property
+    def url(self):
+        """
+        Instance URL
+        """
+        return u'{0.protocol}://{0.domain}/'.format(self)
+
+    @property
+    def active_server_set(self):
+        """
+        Returns the subset of `self.server_set` which aren't terminated
+        """
+        return self.server_set.exclude_terminated()
```

Comment:
@dachary Yup, I'm afraid of these type of structural incompatibilities. However, I've tried to keep an abstraction layer anyway, because we might need to add support for AWS servers - this is what a lot of clients and community members use, and if we ever get to handle the production instances we currently manage with this, I'm not sure they would all agree to move to OpenStack. Having support for both, even if it is tricky, might help to smoothen the migration path toward OpenStack. I am not sure how realistic it is, but it doesn't cost us much to try to keep an abstraction level in the meantime.

@e-kolpakov For `exclude_terminated` specifically, with the abstractions currently in the code, that ought to be a method common to all `Server` models, you are correct. It's actually a method from the queryset/manager, which doesn't have an abstract class currently though - there are other methods on it, like `create()`, that might require to be moved too; I'll have a look to see what could be done simply.